snuba_sdk/metrics_visitors.py

```diff
@@ -229,3 +236,141 @@ def visit(self, scope: MetricsScope) -> str:
+        }
+
+        return ret
+
+    def visit(self, formula: Formula) -> Mapping[str, str]:
+        parameters = []
+        if formula.parameters is not None:
+            parameters = [
+                self._visit_parameter(p, formula.filters) for p in formula.parameters
+            ]
+
+        entity = None
+        for p in parameters:
+            if isinstance(p, dict):
+                if "entity" in p:
+                    if entity is None:
+                        entity = p["entity"]
+                    elif entity != p["entity"]:
+                        raise InvalidExpressionError(
+                            "Formulas can only operate on a single entity"
+                        )
+        assert entity is not None  # This is guaranteed by the Formula validator
+        ret = {"entity": entity}
+
+        # collect all the group bys
+        groupbys = []
+        if formula.groupby:
+            for g in formula.groupby:
+                groupbys.append(self.expression_visitor.visit(g))
+
+        for p in parameters:
+            if isinstance(p, dict) and "groupby" in p:
+                groupbys.append(p["groupby"])
+                break  # The validator guarantees all the groupbys are equal, so we only need one
+
+        ret["groupby"] = ", ".join([g for g in groupbys if g != ""])
+
+        # Collect select statements
+        parameters = ", ".join(p["select"] for p in parameters)
+        ret[
+            "aggregate"
+        ] = f"{formula.operator.value}({parameters}) AS {AGGREGATE_ALIAS}"
+
+        return ret
+
+
+class TimeseriesFormulaSnQLPrinter(TimeseriesVisitor[str]):
+    def __init__(
+        self,
+        expression_visitor: Translation | None = None,
+        metrics_visitor: MetricSnQLPrinter | None = None,
+    ) -> None:
+        self.expression_visitor = expression_visitor or Translation()
+        self.metrics_visitor = metrics_visitor or MetricSnQLPrinter(expression_visitor)
+
+    def _combine(
+        self,
+        timeseries: Timeseries,
+        returns: Mapping[str, str | Mapping[str, str]],
+    ) -> Mapping[str, str]:
+        metric_data = self.metrics_visitor.visit(timeseries.metric)
+        assert isinstance(metric_data, Mapping)  # mypy
+
+        ret = {
+            "entity": metric_data["entity"],
+        }
+
+        metric_filter = Function("equals", [Column("metric_id"), timeseries.metric.id])
+
+        # In Formulas, the filters need to be built into the aggregate as infix functions
+        filters = []
+        if timeseries.filters:
+            for cond in timeseries.filters:
+                assert isinstance(cond.op, Op)  # BooleanOp is not a valid function
+                op = OPERATOR_TO_FUNCTION[cond.op]
+                filters.append(Function(op.value, [cond.lhs, cond.rhs]))
+
+        and_filters = Function("and", [metric_filter, *filters])
+        aggregate = CurriedFunction(
+            f"{timeseries.aggregate}If",
```

Comment:
if we're doing this, should we validate the aggregate to some allowlist? This also opens up the question: which aggregation function do we want to support out the box (if not all) and should they be the same across all entities?

Reply:
I do think we need an allow list, but I would do that in a separate PR from this.